lib/addons/src/transitional-decorator.js

```diff
@@ -0,0 +1,34 @@
+import deprecate from 'util-deprecate';
+
+// Create a decorator that can be used both in the (deprecated) old "hoc" style:
+//   .add('story', decorator(options)(() => <Story />));
+//
+// And in the new, "parameterized" style:
+//   .addDecorator(decorator)
+//   .add('story', () => <Story />, { name: { parameters } });
+
+export const makeTransitionalDecorator = ({ name, parameterName, wrapper }) => {
+  const decorator = options => (getStory, context) => {
+    const parameters = context.parameters && context.parameters[parameterName];
+
+    return wrapper(getStory, context, {
+      options,
+      parameters,
+    });
+  };
+
+  const hoc = deprecate(
+    options => story => context => decorator(options)(story, context),
+    `Passing stories directly into ${name}() is deprecated, instead use addDecorator(${name}) and pass options with the '${parameterName}' parameter`
```

Comment:
Are we sure that we want to deprecate the old way? Honestly, it looks a lot more explicit to me

Reply:
Oh, I thought we did?

I think the strongest argument in favour is if you want to pass parameters to more than one decorator for the same story:

```js
.add('story', withNotes('foo')(backgrounds(['red', 'white'])(() => <Story />)))

vs

.add('story', () => <Story />, { notes: 'foo', backgrounds: ['red', 'white'] })
```

I was planning on changing all the documentation to not mention the old way at least, so deprecating made sense to me, but I'm open to whatever.

Reply:
What is a benefit of having the old API supported?

Reply:
You may want to supply the same parameters (e.g. backgrounds) to multiple stories

And even extract it into variable or module, like here: 
https://github.com/storybooks/storybook/tree/master/addons/backgrounds#usage
("Of course it's easy to create a library module")

Reply:
```
.addDecorator(backgrounds(myBackgrounds))
.add('story', () => <Story />)
.add('other story', () => <OtherStory />)

vs

.add('story', () => <Story />, { backgrounds: myBackgrounds })
.add('otherStory', () => <OtherStory />, { backgrounds: myBackgrounds })
```

Reply:
Or even global `addDecorator(backgrounds(myBackgrounds)` if you want them for all the stories (which is quite often the case)

Reply:
It would be nice if parameters could be used to override common settings, like this:

```
.addDecorator(backgrounds(myCommonBackgrounds))
.add('story', () => <Story />)
.add('other story', () => <OtherStory />)
.add('special story', () => <SpecialStory />, {backgrounds: specialBackgrounds})
```

The implementation could be as simple as
```
const options = {...commonOptions, context.parameters[parameterName]}
```

Reply:
@Hypnosphi you can!

```js
chapter.addParameters({ backgrounds: myBackgrounds })

// or globally
addParameters({ backgrounds: myBackgrounds });
```

Reply:
Passing options to the the decorator (like `.addDecorator(backgrounds(myCommonBackgrounds))`) will pass those options as a separate argument to the the `wrapper` function, at which point the decorator can do what it wants with them.

My gut feeling is it confusing to have both:

```js
.addDecorator(backgrounds(myBackgrounds))

// OR, equivalent
.addDecorator(backgrounds)
.addParameters({ backgrounds: myBackgrounds })
```

As two ways to achieve the same thing, so I lean towards the second as it is more flexible (you can add the decorator at any level of the stack and the parameters at any level also).

Reply:
```
.addDecorator(backgrounds)
.addParameters({ backgrounds: myBackgrounds })
```

This is both more verbose and more magical =(

Reply:
I mean my idea is that you would just add `addDecorators(backgrounds)` once in your `.storybook/config.js` once (equivalent to adding it to `.storybook/addons.js`), and then you would add the parameters to turn the addon on wherever you needed it.

Reply:
OK this looks like a valid option. In this case we need to make sure that each addon is a noop when it lacks its parameters

Reply:
Yeah! I was going to do this per-addon (notes already is a no-op if the params are not set), but we could also make this thing do it automatically. I guess it depends on whether the addon makes sense without params.

Reply:
> I mean my idea is that you would just add addDecorators(backgrounds) once in your .storybook/config.js once (like you need to add it to .storybook/addons.js), and then you would add the parameters to turn the addon on wherever you needed it.

That's perfect I think.

Reply:
@tmeasday can you please do the above in examples-official?

Reply:
@tmeasday I just realised that those are two different usecases

```
// here withSomething is params => (storyFn, context) => output
.addDecorator(withSomething(params))

// here it is params => storyFn => () => output
.add('story', withSomething(params)(storyFn))
```

Actually addons that support both do some checks to figure out which usage is it.

Can we maybe just deprecate the latter but keep the former, because it's quite useful in cases like backgrounds?

Reply:
Oh, now I see that the latter is the thing we're actually deprecating here, good

Reply:
But the former would probably need to be supported when migrating other addons